File: Data Structures & Algorithms/lowest-common-ancestor-of-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # 如果当前节点为空，说明没找到
        if not root:
            return None

        # 如果当前节点就是 p 或 q，直接返回当前节点
        if root == p or root == q:
            return root

        # 去左子树找 p 或 q
        left = self.lowestCommonAncestor(root.left, p, q)

        # 去右子树找 p 或 q
        right = self.lowestCommonAncestor(root.right, p, q)

        # 如果左右两边都找到了，说明当前节点就是最低公共祖先
        if left and right:
            return root

        # 如果只有左边找到了，就把左边的结果传上去
        if left:
            return left

        # 如果只有右边找到了，就把右边的结果传上去
        return right

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return null;
        if (p.val > q.val) {
            // 保证 p.val <= q.val，便于后续情况讨论
            return lowestCommonAncestor(root, q, p);
        }
        if (root.val >= p.val && root.val <= q.val) {
            // p <= root <= q
            // 即 p 和 q 分别在 root 的左右子树，那么 root 就是 LCA
            return root;
        }
        if (root.val > q.val) {
            // p 和 q 都在 root 的左子树，那么 LCA 在左子树
            return lowestCommonAncestor(root.left, p, q);
        } else {
            // p 和 q 都在 root 的右子树，那么 LCA 在右子树
            return lowestCommonAncestor(root.right, p, q);
        }        
    }
}//ologn,on
